File: Python/lab5/solutions/high_throughput_producer.py
import random
import sys
import time
from kafka import KafkaProducer
import logging

def on_send_success(record_metadata):
    logging.info(record_metadata.topic)
    logging.info(record_metadata.partition)
    logging.info(record_metadata.offset)

def on_send_error(excp):
    logging.error('I am an errback', exc_info=excp)
    # handle exception

producer = KafkaProducer(bootstrap_servers=['localhost:9092', 'localhost:9093'], 
                        acks = 0,
                        batch_size = 16824,
                        linger_ms = 5,
                        max_in_flight_requests_per_connection = 5)
topic = "throughput"
start_time = time.time()
count = 0
while count < 100000:
    producer.send(topic=topic, key=b'1', value=b'1')
    count += 1
producer.flush()
producer.close()
print(f'I have produced {count} messages in {(time.time() - start_time)} seconds')

File: JS/lab5/solutions/high_throughput_producer.js
const { Kafka } = require('kafkajs')
const { performance } = require('perf_hooks');

const kafka = new Kafka({
  brokers: ['localhost:9092', 'localhost:9093'],
  clientId: 'lab4-producer',
})


const producer = kafka.producer({maxInFlightRequests : 5})
const topic = 'throughput' 
const sendMessage = () => {
  var messages = [];
  for (let i = 0; i < 1000; i++) {
    messages.push({key: '1', value: '1'})
  }
  return producer.send({
      topic,
      messages: messages,
      acks: 0
  })
}

const run = async () => {
  await producer.connect()
  var startTime = performance.now()
  var count = 0
  while (performance.now() - startTime < 10000) {
    await sendMessage().catch(e => console.error(`[example/producer] ${e.message}`, e));
    count++
  }
  console.log(`I have sent ${count*1000} messages in 10 seconds`)
}
run()